### ISSUE TITLE
[Feature request] Add auxiliar support to the VJP

### ISSUE THREAD

**Issue by abdussamettrkr** [ml-explore/mlx#644]
Adding support for auxiliary outputs to the VJP could be quite useful in some cases. Users may not want to differentiate all outputs which is currently not supported directly. Something similar to Jax and Torch could be beneficial. 

Jax and Torch have an additional flag, 'has_aux,' which, when set to false, expects the function to return one output; otherwise, it returns two outputs.

**Comment by awni** [comment]
You don't need `has_aux` in MLX, we just return the full output in the value spot:

```
import mlx.core as mx

def fun(x):
    return x.sum(), x + 1, x + 2, x + 3

(loss, v1, v2, v3), grad = mx.value_and_grad(fun)(mx.array(2.0))
```

**Comment by abdussamettrkr** [comment]
I see, but in my case I need to use VJP to pass cotangent vector. Therefore, similar cases to me may need similar VJP implementation like 'mx.value_and_grad'.

**Comment by awni** [comment]
Example?

**Comment by abdussamettrkr** [comment]
I created a simple example below. Thank you for interest.

```python
import jax
import jax.numpy as jnp

def func(arr1: jnp.array, arr2: jnp.array):
    out1 = arr1 + arr2*2
    out2 = (
        arr1*2 + arr2*2
    )  # My functions just calculates out2, don't want to effect grads
    return out1, out2

inp1 = jnp.array([4., 8, 16])
inp2 = jnp.array([4., 5, 6])
cotangents = jnp.array([22., 10, 82])

out, vjp_fn, aux = jax.vjp(func, inp1, inp2, has_aux=True)
grads = vjp_fn(cotangents)
print("Grads:", grads)
```

Prints:
Grads: (Array([22., 10., 82.], dtype=float32), Array([ 44.,  20., 164.], dtype=float32))

**Comment by awni** [comment]
Got it. I can see how that would be convenient.

I'm not a huge fan of `has_aux`, but maybe we could allow you to specify the VJP outputs in the transform itself.. 🤔

**Comment by angeloskath** [comment]
There is no need for `has_aux` you just need to wrap `out2` with `stop_gradient`. If you don't require gradient for an output then you don't need to pass a cotangent. I think this is actually more intuitive than `has_aux`. For instance:

```python
import mlx.core as mx

def func(arr1, arr2):
    out1 = arr1 + arr2*2
    out2 = arr1*2 + arr2*2
    return out1, mx.stop_gradient(out2)

inp1 = mx.array([4., 8, 16])
inp2 = mx.array([4., 5, 6])
cotangents = mx.array([22., 10, 82])
values, grads = mx.vjp(func, [inp1, inp2], [cotangents])
print(grads)
# [array([22, 10, 82], dtype=float32), array([44, 20, 164], dtype=float32)]
```

**Comment by abdussamettrkr** [comment]
Yes, I see. I think that will work, we don't need has_aux.

**Comment by angeloskath** [comment]
Sounds good. I will close the issue but feel free to reopen if you find anything inconvenient in the vjp transform.
